src/mem/mem.c

```diff
@@ -29,9 +29,9 @@
 struct mem {
 	size_t nrefs;          /**< Number of references  */
 	mem_destroy_h *dh;     /**< Destroy handler       */
+	size_t size;           /**< Size of memory object */
```

Comment:
is this change really necessary ?

it will add 8 bytes overhead to all allocations, in RELEASE build

Reply:
`size` is needed for the `mem_realloc` changes. If we're dropping it, this change will also go away.

PS: It would increase the header size by 16 bytes doe to alignment. But it would unlikely matter since `malloc` also aligns its allocation sizes for binning.

Reply:
Actually, no, `size` is still needed for `mem_realloc` even if the optimization is removed in order to fix the correctness issue when `nrefs` > 1. `mem_realloc` needs `size` in order to copy data to the newly allocated segment.

Reply:
BTW, if you want `mem` to fit in 16 bytes on 64-bit targets (instead of 32), you could change `nrefs` and `size` to `uint32_t`. This would limit mem allocation sizes to 4 GiB max, but that doesn't seem like a significant limitation for libre use cases.

Reply:
Using `uint32_t` for `nrefs` and `size` looks fine to me. The 8 bytes overhead on 32-bit systems remains (using 12 byte and aligned/padded to 16 byte in size). But I see no better solution if we fix `mem_realloc` with `mem_alloc` and copy. If we return `NULL`, we have to fix for example `mbuf_resize` if `mbuf_alloc_ref` is used. This is a serious issue too since as lastique mentioned `realloc` will call `free` if the pointed area was moved and any referenced `mbuf->buf` pointers are dangling. And its very likely that mbuf->buf is changed with realloc. 

@alfredh let me know if you see a better option to fix this. I will test this patch a bit more on multiple systems, but I think  the 8 byte overhead on 32-bit is something we have to accept to fix mem_realloc.

And maybe we can reuse the mem->size for other productive uses-cases in mem. For example I experimented with a basic/optional `out-of-bound` check, but was not sure how helpful/useful it would be.

Reply:
I've updated the PR to switch `nrefs` and `size` to `uint32_t`.